app/src/main/java/tech/relaycorp/gateway/domain/publicsync/RegisterGateway.kt

```diff
@@ -0,0 +1,45 @@
+package tech.relaycorp.gateway.domain.publicsync
+
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.data.model.RegistrationState
+import tech.relaycorp.gateway.data.preference.PublicGatewayPreferences
+import tech.relaycorp.gateway.domain.LocalConfig
+import tech.relaycorp.poweb.PoWebClient
+import tech.relaycorp.relaynet.messages.control.PrivateNodeRegistration
+import java.net.URL
+import java.util.logging.Level
+import javax.inject.Inject
+
+@JvmSuppressWildcards
+class RegisterGateway
+@Inject constructor(
+    private val publicGatewayPreferences: PublicGatewayPreferences,
+    private val localConfig: LocalConfig,
+    private val poWebClientBuilder: ((String) -> PoWebClient)
+) {
+
+    suspend fun registerIfNeeded() {
+        if (publicGatewayPreferences.getRegistrationState() != RegistrationState.ToDo) {
+            return
+        }
+
+        val pnr = register() ?: return
+        localConfig.setCertificate(pnr.privateNodeCertificate)
+        publicGatewayPreferences.setCertificate(pnr.gatewayCertificate)
+        publicGatewayPreferences.setRegistrationState(RegistrationState.Done)
+    }
+
+    private suspend fun register(): PrivateNodeRegistration? =
+        try {
+            val address = publicGatewayPreferences.getAddress()
+            val hostName = URL(address).host
+            val poWeb = poWebClientBuilder.invoke(hostName)
+            val keyPair = localConfig.getKeyPair()
+
+            val pnrr = poWeb.preRegisterNode(keyPair.public)
+            poWeb.registerNode(pnrr.serialize(keyPair.private))
+        } catch (e: Exception) {
```

Comment:
I think it'd be better to catch a more concrete class, like `PDCException`. Or even better: even split this into two, like this...

- `ServerException` -> log a debug entry.
- `ClientBindingException` -> log an error. That means that the Android Gateway or the PoWeb library is not complying with the specs.